ppi.py

```diff
@@ -0,0 +1,89 @@
+from binaryninja import BinaryView, MediumLevelILCall, MediumLevelILTailcall, MediumLevelILLoadStruct, PointerType, \
+    MediumLevelILConstPtr, BackgroundTask
+from .protocols import lookup_protocol_guid, define_protocol_types
+from .system_table import propagate_system_table_pointers
+
+
+def define_descriptor(bv: BinaryView, task):
+    descriptor_types = ["EFI_PEI_NOTIFY_DESCRIPTOR", "EFI_PEI_PPI_DESCRIPTOR", "EFI_PEI_DESCRIPTOR", ]
+    for descriptor_type in descriptor_types:
+        refs = list(bv.get_code_refs_for_type(descriptor_type))
+        for ref in refs:
+            if task.cancelled:
+                return False
+            instr = ref.mlil
+            if isinstance(instr, MediumLevelILCall) or isinstance(instr, MediumLevelILTailcall):
+                if isinstance(instr.dest, MediumLevelILLoadStruct):
+                    # likely a call to services
+                    if isinstance(instr.dest.expr_type, PointerType):
+                        function_type = instr.dest.expr_type.target
```

Comment:
There is a lot of code in this block. I recommend creating a function to make it more readable.